### PR TITLE
Add nDC support for workflow execution reset

### PR DESCRIPTION
**Pull request by wxing1292**
* New nDCWorkflowResetter handling workflow execution reset
* Add handling of workflow reset in nDCHistoryReplicator
* Add ContainsItem API to version history
* Rename nDCStateRebuilder to nDCConflictResolver
* Move common mutable state rebuild logic into new nDCStateRebuilder
* Add '+build test' for all NDC related test files
* Add //go:generate mockgen for automated mock file generation

### CODE REVIEW

File: service/history/nDCBranchMgr_mock.go [hunk 1]
```diff
@@ -1,4 +1,6 @@
-// Copyright (c) 2019 Uber Technologies, Inc.
+// The MIT License (MIT)
+//
+// Copyright (c) 2017 Uber Technologies, Inc.
```
Comment (yycptt) [inline service/history/nDCBranchMgr_mock.go]: Update the LICENSE file?

File: service/history/nDCBranchMgr.go [hunk 2]
```diff
@@ -18,6 +18,8 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
+//go:generate mockgen -copyright_file ../../LICENSE -package $GOPACKAGE -source $GOFILE -destination nDCBranchMgr_mock.go
```
Comment (yycptt) [inline service/history/nDCBranchMgr.go]: Add `go generate` to Makefile?

File: service/history/nDCBranchMgr_mock.go (hunk 1 above)
Comment (wxing1292) [reply service/history/nDCBranchMgr_mock.go]: this is the new license file

File: service/history/nDCBranchMgr.go (hunk 2 above)
Comment (wxing1292) [reply service/history/nDCBranchMgr.go]: not really, this should be generated manually

if interface is changed, then the test should not work at all

File: service/history/nDCBranchMgr_mock.go (hunk 1 above)
Comment (yycptt) [reply service/history/nDCBranchMgr_mock.go]: I mean we should change 2017 -> 2019 in the LICENSE file. Maybe in another diff.

File: service/history/nDCStateRebuilder_test.go [hunk 3]
```diff
@@ -28,13 +30,15 @@ import (
 	"github.com/stretchr/testify/mock"
 	"github.com/stretchr/testify/suite"
 	"github.com/uber-go/tally"
+
```
Comment (shreyassrivatsan) [inline service/history/nDCStateRebuilder_test.go]: nit: spurious newline.

File: service/history/nDCWorkflowResetter.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+//go:generate mockgen -copyright_file ../../LICENSE -package $GOPACKAGE -source $GOFILE -destination nDCWorkflowResetter_mock.go
+
+package history
+
+import (
+	ctx "context"
+
+	"github.com/pborman/uuid"
+
```
Comment (shreyassrivatsan) [inline service/history/nDCWorkflowResetter.go]: spurious newline.

File: service/history/nDCWorkflowResetter.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+		logger log.Logger
+	}
+)
+
+var _ nDCWorkflowResetter = (*nDCWorkflowResetterImpl)(nil)
+
+func newNDCWorkflowResetter(
+	shard ShardContext,
+	transactionMgr nDCTransactionMgr,
+
```
Comment (shreyassrivatsan) [inline service/history/nDCWorkflowResetter.go]: spurious newline.

File: service/history/nDCBranchMgr.go [hunk 6]
```diff
@@ -139,7 +141,7 @@ func (r *nDCBranchMgrImpl) createNewBranch(
 	defer func() {
 		if errComplete := r.historyV2Mgr.CompleteForkBranch(&persistence.CompleteForkBranchRequest{
 			BranchToken: newBranchToken,
-			Success:     retError == nil || persistence.IsTimeoutError(retError),
+			Success:     true, // past lessons learnt from Cassandra & gocql tells that we cannot possibly find all timeout errors
```
Comment (vancexu) [inline service/history/nDCBranchMgr.go]: What is it means?

File: service/history/nDCWorkflowResetter.go [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+		r.baseRunID,
+	)
+	resetWorkflowIdentifier := definition.NewWorkflowIdentifier(
+		r.domainID,
+		r.workflowID,
+		r.newRunID,
+	)
+
+	requestID := uuid.New()
+	rebuildMutableState, rebuildHistorySize, err := r.stateRebuilder.rebuild(
```
Comment (shreyassrivatsan) [inline service/history/nDCWorkflowResetter.go]: use `rebuilt` here?

File: service/history/nDCBranchMgr.go (hunk 6 above)
Comment (wxing1292) [reply service/history/nDCBranchMgr.go]: this means that basically for all new branch created, the logic will mark it as success.
this is due to lessons learnt from using Cassandra & gocql, e.g. a error a timeout, or '*only receive 1 / 2 response' error, or other errors, does not really means that the query issued to Cassandra will fail.

the only safe thing to do it mark the branch creation as success, and tolerate data leak, until our background cleanup logic is implemented

File: service/history/nDCWorkflowResetter.go (hunk 5 above)
Comment (wxing1292) [reply service/history/nDCWorkflowResetter.go]: this is intentional, making the input variables easy to read

**Pull request merged**